src/testrender/sampling.h

```diff
@@ -14,18 +14,28 @@ OSL_NAMESPACE_ENTER
 
 struct TangentFrame {
     // build frame from unit normal
-    TangentFrame(const Vec3& n) : w(n)
-    {
-        u = (fabsf(w.x) > .01f ? Vec3(w.z, 0, -w.x) : Vec3(0, -w.z, w.y))
-                .normalize();
-        v = w.cross(u);
+    static TangentFrame from_normal(const Vec3& n) {
+        // https://graphics.pixar.com/library/OrthonormalB/paper.pdf
+        const float sign = copysignf(1.0f, n.z);
+        const float a = -1 / (sign + n.z);
+        const float b = n.x * n.y * a;
+        const Vec3 u = Vec3(1 + sign * n.x * n.x * a, sign * b, -sign * n.x);
+        const Vec3 v = Vec3(b, sign + n.y * n.y * a, -n.y);
+        return { u, v, n };
     }
 
     // build frame from unit normal and unit tangent
-    TangentFrame(const Vec3& n, const Vec3& t) : w(n)
-    {
-        v = w.cross(t);
-        u = v.cross(w);
+    // fallsback to an arbitrary basis if the tangent is 0 or colinear with n
+    static TangentFrame from_normal_and_tangent(const Vec3& n, const Vec3& t) {
+        Vec3 y = n.cross(t);
```

Comment:
Out of curiosity, was there a reason you went with computing y here rather than x (via t - n*dot(n, t))? Numerical stability or performance?

Reply:
No real reason actually. Now that I look at both in the light of day, they should be equivalent, and the dot product way might be a few less operations (especially considering the dot product is needed for the LTC lookup anyway).

Reply:
Okay, fair enough. I just wanted to check that I wasn't missing anything.